refactor(esp_tinyusb): Run-time configuration for task and peripheral port

roma-jam opened the pull request:
## Description

This is the first MR in the sequence which contains the run-time configuration for TinyUSB.
 
## Related

- Run-time TinyUSB configuration

## Testing
- Device: All configurations, present and default (when not provided or NULL)
- Peripheral: Default periph + Periph selection (Full- or High-speed) when available
- Task: Default task with different affinity mask 

<!--
## Memory usage of libesp_tinyusb.a (esp-idf latest)

| Memory | Buildtime FS | Buildtime FS & HS |  Runtime FS only | Runtime FS & HS |
| ------------- | ------------- | --------------- |  ------------| ---- | 
| DIRAM   |  142 |154 | 166 |  178|    
|  .bss |  116 | 128 | 132 |  144|   
| .data |  26 | 26 | 34 | 34 | 
| .text |  0 | 0 | 0 | 0 | 
| Flash |  1563 | 3380| 2720 |  4912 | 
| Total | 2046 |3534 | 3239 |  5090 |  
-->

---

## Checklist

Before submitting a Pull Request, please ensure the following:

- [ ] 🚨 This PR does not introduce breaking changes.
- [x] All CI checks (GH Actions) pass.
- [ ] Documentation is updated as needed.
- [x] Tests are updated or added as necessary.
- [x] Code is well-commented, especially in complex areas.
- [x] Git history is clean — commits are squashed to the minimum necessary.

peter-marcisovsky commented on device/esp_tinyusb/tinyusb_task.c:
```diff
@@ -0,0 +1,215 @@
+
+    if (task_cfg->init_in_task) {
+        // Wait until tinyusb task has started and complete initialization of the driver stack
+        EventBits_t flags = xEventGroupWaitBits(p_tusb_task_ctx->flags,
```
Can you, for example, use `xTaskNotifyGiveIndexed()` here, instead of the event group wait bits? So you don't have to deal with event group create and delete.

Or just the normal `xTaskNotifyGive()`, which you will give in the `tinyusb_device_task()` regardless the `tusb_rhport_init()` failed or not. And afterwards, back in the `tinyusb_task_start`, once `ulTaskNotifyTake()`, you will check whether the `tinyusb_device_task()` is created, which will determine whether the `tusb_rhport_init` has failed or not.

peter-marcisovsky commented on device/esp_tinyusb/tinyusb_task.c:
```diff
@@ -0,0 +1,215 @@
+        TINYUSB_TASK_EXIT_CRITICAL();
+        ESP_LOGE(TAG, "TinyUSB task has already started");
+        ret = ESP_ERR_INVALID_STATE;
+        goto assign_err;
```
Once you hit this goto, it will also do the `vEventGroupDelete`, which you already did couple of lines earlier.

peter-marcisovsky commented on device/esp_tinyusb/include/tinyusb_task.h:
```diff
@@ -0,0 +1,89 @@
+ *    - ESP_ERR_INVALID_ARG if task_cfg is NULL or init_params is NULL when init_in_task is true
+ *    - ESP_ERR_NOT_FINISHED if TinyUSB task creation failed
+ *    - ESP_ERR_NO_MEM if memory allocation failed
+ *    - ESP_OK if TinyUSB Task initialized successfully
```
`ESP_ERR_INVALID_STATE` retval description is missing

peter-marcisovsky commented on device/esp_tinyusb/tinyusb_task.c:
```diff
@@ -0,0 +1,215 @@
+task_err:
+    if (task_cfg->init_in_task) {
+        vEventGroupDelete(p_tusb_task_ctx->flags);
+    }
```
Somewhere here, you should also set `p_tusb_task_ctx = NULL`

If one installation fails after you do `p_tusb_task_ctx = task_ctx`, then you will not be able to call the `tinyusb_task_start()` again.

peter-marcisovsky commented on device/esp_tinyusb/tinyusb_task.c:
```diff
@@ -0,0 +1,215 @@
+/*
+ * SPDX-FileCopyrightText: 2020-2025 Espressif Systems (Shanghai) CO LTD
```
Also a new file, license year should be a current year.

peter-marcisovsky reviewed (commented):
Before I get digging into the test app, can you please provide some summary about the test app refactor?

Did you change something inside the test app? Aka.. added test cases, removed test cases. Or is it just a refactor in terms of files rearrangement?

roma-jam commented:
@peter-marcisovsky ,

thanks a lot for the prompt review! 

Test applications:
- I removed the test apps, related to task and now they are present (with some additional test cases) in the `test_task.c` file.
- I moved configuration tests to the `test_device.c` fully, without any changes on the logic. 
- I added `test_periph.c` which holds the cases with explicit configuration for the perhipheral.

So, I should not: 
- decrease the coverage with refactoring
- add more specific cases and scenarios, because of adding the potential runtime configuration
- removed the separate test applications, which now are a part of runtime configuration test. 


Please, let me know if you need more information.

peter-marcisovsky commented on device/esp_tinyusb/test_apps/runtime_config/main/test_device.c:
```diff
@@ -0,0 +1,406 @@
+ * Scenario: Provide maximum supported string descriptors
+ * Awaiting: Install returns ESP_OK, device is enumerated, tusb_mount_cb() is called
+ */
+TEST_CASE("Device: String Descriptos maximum value", "[runtime_config][default]")
```
A typo here: 

```suggestion
TEST_CASE("Device: String Descriptors maximum value", "[runtime_config][default]")
```

roma-jam commented:
@peter-marcisovsky @tore-espressif 
I have updated the PR, PTAL. 

Tests are also available for runtime_config, so check them first to see how the new configuration mechanism is working. 

Thanks.

tore-espressif reviewed (approved):
@roma-jam I left comments that could be improved. And 2 nitpicks that I leave for your decision.

Overall, LGTM!

peter-marcisovsky commented on device/esp_tinyusb/test_apps/runtime_config/pytest_runtime_config.py:
```diff
@@ -0,0 +1,22 @@
+    ]
+
+    for periph in peripherals:
+        dut.run_all_single_board_cases(group=periph, reset=True)
```
Do you need to reset the DUT after each test case? If not, you can remove the `reset=True`.
We are using `reset=True` in esp-idf for host tests only, where it helps to pass certain test cases.

roma-jam replied:
No, I don't think so. 
Thanks, removed the reset.

peter-marcisovsky reviewed (approved):
Thanks for the update. LGTM.

roma-jam commented on host/class/cdc/usb_host_cdc_acm/test_app/main/usb_device.c:
```diff
@@ -66,23 +67,19 @@ static const tusb_desc_device_qualifier_t device_qualifier = {
-        .device_descriptor = &cdc_device_descriptor,
-        .string_descriptor = NULL,
-        .string_descriptor_count = 0,
+    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();
```
Currently, default config covers only phy, peripheral port number and task configuration. 

Descriptors configuration is not covered in this MR. 

So, speaking of advantage: just not to think about default parameters. If we will change them later, we don't need to touch tests source code, just default macros.

roma-jam replied:
I have added the test "Multitask: Install" when I tried to install the driver from several tasks, and found a bug with descriptor configuration: descriptor configuration corrupted, when another install is called, when skip_phy_setup is true.

ATM the test is not running. 

It is better to fix it at the same moment with refactoring the descriptor configuration, so I'll address this in upcoming PR.

roma-jam commented on device/esp_tinyusb/tinyusb_task.c:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    // Wait until the port will be active, 5 sec is more than enough
+    if (xSemaphoreTake(task_ctx->port_active_smp, pdMS_TO_TICKS(5000)) == pdFALSE) {
```
Thanks for the suggestion, changed to `xTaskNotifyGive`.

roma-jam commented on device/esp_tinyusb/include/tinyusb_default_configs.h:
```diff
@@ -0,0 +1,85 @@
+ *
+ * Default task affinity:
+ * - ESP32P4:       CPU1
+ * - ESP32S2/S3:    CPU0
```
because I did split it by speed, not by UNICORE option. 
Now it is better, thanks.

roma-jam commented on device/esp_tinyusb/test_apps/runtime_config/main/test_device.c:
```diff
@@ -0,0 +1,299 @@
+
+#if (TUD_OPT_HIGH_SPEED)
+
+static const uint8_t test_hs_configuration_descriptor[] = {
```
Thanks, done.

roma-jam commented on device/esp_tinyusb/test_apps/runtime_config/main/test_device.c:
```diff
@@ -0,0 +1,299 @@
+/**
+ * @brief TinyUSB Task specific testcase
+ *
+ * Scenario: Provide only device descriptor
```
Yes, thanks for the note.

roma-jam commented on device/esp_tinyusb/include/tinyusb_task.h:
```diff
@@ -0,0 +1,68 @@
+ *    - ESP_ERR_INVALID_ARG if task_cfg is NULL or init_params is NULL when init_in_task is true
+ *    - ESP_ERR_NOT_FINISHED if TinyUSB task creation failed
+ *    - ESP_ERR_NO_MEM if memory allocation failed
+ *    - ESP_OK if TinyUSB Task initialized successfully
```
Added it, thanks.
